presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -230,6 +231,11 @@ private void processNewSources()
 
         requireNonNull(duration, "duration is null");
 
+        // if the driver is blocked we don't need to continue
+        if (!previouslyReturnedBlockedFuture.isDone()) {
```

Comment:
do we need synchronization for `previouslyReturnedBlockedFuture`?

Reply:
Is it almost the same question as "do we need this condition at all?" 
It looks like both have same answer: "not needed" -- at least at the first sight. But I heard there's something more to it.

However, if youj move the condition inside `tryWithLock` block, then all doubts seem to disappear.

Reply:
Then we also need to synchronize the writes to this variable with the `exclusiveLock` and things get messy easily.

Reply:
Yup, write is already (hopefully) well guarded